File: src/core/engine/pluginProps.ts
/**
 * PLUGIN properties (B3z, docs/ENGINE_API.md §15.9 "Plugin properties") — the
 * values JavaScript plugins keep on a layer, as ordinary engine-API properties.
 *
 *   plugin/<name>                  a plugin-declared LAYER KIND's prop (the
 *                                  layer's `pluginLayer:<…>` component; keyed on
 *                                  the `plugin.<name>` track — customLayers.ts)
 *   plugin/<slug>/<panel>          the property GROUP of one contributed
 *                                  inspector panel (the `PluginParams.<slug>.<panel>`
 *                                  component, id `pluginui_<slug>_<panel>` —
 *                                  uiParams.ts); added / removed with
 *                                  `addPropertyGroup` / `removePropertyGroups`
 *   plugin/<slug>/<panel>/<name>   one of its params (keyed on the
 *                                  `pluginUi.<slug>.<panel>.<name>[.<axis>]` tracks)
 *
 * The engine has no plugin SCHEMA: a plugin's declaration is editor state
 * (the installed manifest), not document state, and the C++ engine never sees
 * it. So a binding is typed by what the DOCUMENT stores — After Effects' model
 * for an effect instance whose plug-in is missing: the params are still there
 * and editable by type. A number is an animatable scalar (a point's `<name>.x` /
 * `.y` / `.z` numbers one vec2 / vec3 — the AE Point Control); a boolean a bool,
 * a string a string (an enum value, a colour hex, an asset id), anything else
 * (null, objects) json. A non-numeric param also takes ANY json value (the
 * plug-in's arbitrary data: an asset slot goes from null to an id). A panel
 * group is added with ALL its declared defaults (`init` — the client knows the
 * schema), exactly what the editor's first write used to seed.
 *
 * Bindings are added in a fixed, SORTED order (property names, then panel
 * component types) so both engines list the same tree whatever order the
 * document's objects carry their keys in. Pure over the document; the C++
 * engine ports this file (native/engine/src/core/plugin_props.cpp).
 */

import type { Value } from '@motion/engine-api';
import type { SceneNode } from '@core/types';
import defaultSceneGraph from '@core/scene/DefaultSceneGraph';
import { channelsToColor } from '@core/effects/effects';
import { fail } from './errors';
import type { PropBinding } from './props';

/** customLayers.ts COMPONENT_TYPE_PREFIX / CUSTOM_PROP_PREFIX. */
export const PLUGIN_LAYER_COMPONENT_PREFIX = 'pluginLayer:';
export const PLUGIN_LAYER_TRACK_PREFIX = 'plugin.';
/** uiParams.ts pluginParamComponentType / PLUGIN_PARAM_PREFIX / pluginParamComponentId. */
export const PLUGIN_PANEL_COMPONENT_PREFIX = 'PluginParams.';
export const PLUGIN_PANEL_TRACK_PREFIX = 'pluginUi.';
export const PLUGIN_PANEL_ID_PREFIX = 'pluginui_';

const AXES = ['x', 'y', 'z'] as const;
const SEG = /^[A-Za-z0-9-]+$/;

export type PluginTrack =
  | { kind: 'layer'; name: string }
  | { kind: 'panel'; slug: string; panel: string; name: string; axis?: string };

/** `plugin.<name>` / `pluginUi.<slug>.<panel>.<name>[.<axis>]`, or null. */
export function parsePluginTrack(prop: string): PluginTrack | null {
  if (prop.startsWith(PLUGIN_PANEL_TRACK_PREFIX)) {
    const seg = prop.slice(PLUGIN_PANEL_TRACK_PREFIX.length).split('.');
    if (seg.length < 3 || seg.length > 4 || !seg.every((s) => SEG.test(s))) return null;
    const axis = seg[3];
    if (axis !== undefined && !(AXES as readonly string[]).includes(axis)) return null;
    return { kind: 'panel', slug: seg[0]!, panel: seg[1]!, name: seg[2]!, ...(axis !== undefined ? { axis } : {}) };
  }
  if (prop.startsWith(PLUGIN_LAYER_TRACK_PREFIX)) {
    const name = prop.slice(PLUGIN_LAYER_TRACK_PREFIX.length);
    return SEG.test(name) && !name.startsWith('_') ? { kind: 'layer', name } : null;
  }
  return null;
}

/** The API path of a plugin member track (apiPathFor), or null. */
export function pluginApiPath(prop: string): string | null {
  const t = parsePluginTrack(prop);
  if (!t) return null;
  return t.kind === 'layer' ? `plugin/${t.name}` : `plugin/${t.slug}/${t.panel}/${t.name}`;
}

type Comp = SceneNode['components'][number];

function layerKindComponent(node: SceneNode): Comp | undefined {
  return node.components.find((c) => c.type.startsWith(PLUGIN_LAYER_COMPONENT_PREFIX));
}

function panelComponent(node: SceneNode, slug: string, panel: string): Comp | undefined {
  const type = `${PLUGIN_PANEL_COMPONENT_PREFIX}${slug}.${panel}`;
  return node.components.find((c) => c.type === type);
}

/** `PluginParams.<slug>.<panel>` → {slug, panel}, or null. */
function parsePanelType(type: string): { slug: string; panel: string } | null {
  if (!type.startsWith(PLUGIN_PANEL_COMPONENT_PREFIX)) return null;
  const seg = type.slice(PLUGIN_PANEL_COMPONENT_PREFIX.length).split('.');
  if (seg.length !== 2 || !seg.every((s) => SEG.test(s))) return null;
  return { slug: seg[0]!, panel: seg[1]! };
}

/** A stored key that is a plugin's own (not the host's `__kind`, `__cid`, …). */
const ownKey = (k: string): boolean => !k.startsWith('_') && SEG.test(k.split('.')[0]!) && k.split('.').length <= 2;

// ── The static seam (numbers) ─────────────────────────────────────────

/** The component + key a plugin member track's static value lives at, or null. */
function staticHome(node: SceneNode, t: PluginTrack): { comp: Comp | undefined; key: string } {
  if (t.kind === 'layer') return { comp: layerKindComponent(node), key: t.name };
  return { comp: panelComponent(node, t.slug, t.panel), key: t.axis ? `${t.name}.${t.axis}` : t.name };
}

/**
 * propertyValue.ts seam: the static number of a plugin member track —
 * `undefined` when it stores none, `null` when `prop` is not a plugin track.
 */
export function readPluginStatic(node: SceneNode, prop: string): number | undefined | null {
  const t = parsePluginTrack(prop);
  if (!t) return null;
  const { comp, key } = staticHome(node, t);
  const v = (comp?.props as Record<string, unknown> | undefined)?.[key];
  return typeof v === 'number' ? v : undefined;
}

/** Write it: false when the layer has no such component (nothing is created); null = not a plugin track. */
export function writePluginStatic(nodeId: string, node: SceneNode, prop: string, value: number): boolean | null {
  const t = parsePluginTrack(prop);
  if (!t) return null;
  const { comp, key } = staticHome(node, t);
  if (!comp) return false;
  defaultSceneGraph.writeProp(nodeId, comp.id, key, value);
  return true;
}

// ── Bindings ─────────────────────────────────────────────────────────

/** A non-numeric stored value's type. */
function fieldType(v: unknown): 'bool' | 'string' | 'json' {
  return typeof v === 'boolean' ? 'bool' : typeof v === 'string' ? 'string' : 'json';
}

function numericBinding(path: string, name: string, members: string[]): PropBinding {
  const vt = members.length <= 1 ? 'scalar' : members.length === 2 ? 'vec2' : 'vec3';
  // `home: []`: the static seam owns the value (a write with no component to
  // hold it is `notFound` — nothing is invented on the Transform).
  return { path, name, matchName: members[0]!, valueType: vt, members, animatable: true, unit: '', home: [] };
}

function fieldBinding(path: string, name: string, component: string, key: string, stored: unknown): PropBinding {
  return {
    path, name, matchName: key, valueType: fieldType(stored), members: [], special: 'field',
    field: { owner: 'plugin', key, groupId: component }, animatable: false, unit: '',
  };
}

/**
 * Add the plugin bindings of a layer (props.ts catalogFor, after the latent
 * block): the layer kind's props, then each panel's params — numbers from the
 * stored values AND from animated tracks with no stored value.
 */
export function addPluginBindings(node: SceneNode, trackNames: readonly string[], add: (b: PropBinding) => void): void {
  const tracks = trackNames.map((p) => ({ p, t: parsePluginTrack(p) })).filter((x) => x.t !== null) as Array<{ p: string; t: PluginTrack }>;
  const kind = layerKindComponent(node);
  if (kind) {
    const props = kind.props as Record<string, unknown>;
    const names = new Set<string>();
    for (const k of Object.keys(props)) if (ownKey(k) && !k.includes('.')) names.add(k);
    for (const { t } of tracks) if (t.kind === 'layer') names.add(t.name);
    for (const name of [...names].sort()) {
      const v = props[name];
      const path = `plugin/${name}`;
      if (v === undefined || typeof v === 'number') add(numericBinding(path, name, [`${PLUGIN_LAYER_TRACK_PREFIX}${name}`]));
      else add(fieldBinding(path, name, kind.type, name, v));
    }
  }
  // Panels: the stored components, plus panels only animated tracks name.
  const panels = new Map<string, { slug: string; panel: string; comp?: Comp }>();
  for (const c of node.components) {
    const p = parsePanelType(c.type);
    if (p) panels.set(c.type, { ...p, comp: c });
  }
  for (const { t } of tracks) {
    if (t.kind !== 'panel') continue;
    const type = `${PLUGIN_PANEL_COMPONENT_PREFIX}${t.slug}.${t.panel}`;
    if (!panels.has(type)) panels.set(type, { slug: t.slug, panel: t.panel });
  }
  for (const type of [...panels.keys()].sort()) {
    const { slug, panel, comp } = panels.get(type)!;
    const props = (comp?.props ?? {}) as Record<string, unknown>;
    // name → the axes a number is stored / animated on ('' = a plain scalar).
    const numeric = new Map<string, Set<string>>();
    const other = new Map<string, unknown>();
    for (const [k, v] of Object.entries(props)) {
      if (!ownKey(k)) continue;
      const [name, axis] = k.split('.') as [string, string | undefined];
      if (typeof v === 'number') {
        if (axis !== undefined && !(AXES as readonly string[]).includes(axis)) continue;
        if (!numeric.has(name)) numeric.set(name, new Set());
        numeric.get(name)!.add(axis ?? '');
      } else if (axis === undefined) {
        other.set(name, v);
      }
    }
    for (const { t } of tracks) {
      if (t.kind !== 'panel' || t.slug !== slug || t.panel !== panel || other.has(t.name)) continue;
      if (!numeric.has(t.name)) numeric.set(t.name, new Set());
      numeric.get(t.name)!.add(t.axis ?? '');
    }
    const base = `plugin/${slug}/${panel}`;
    const prefix = `${PLUGIN_PANEL_TRACK_PREFIX}${slug}.${panel}.`;
    for (const name of [...new Set([...numeric.keys(), ...other.keys()])].sort()) {
      const path = `${base}/${name}`;
      if (other.has(name)) {
        add(fieldBinding(path, name, type, name, other.get(name)));
        continue;
      }
      const axes = AXES.filter((a) => numeric.get(name)!.has(a));
      const members = axes.length > 0 ? axes.map((a) => `${prefix}${name}.${a}`) : [`${prefix}${name}`];
      add(numericBinding(path, name, members));
    }
  }
}

/** The panel group paths of a layer (`plugin/<slug>/<panel>`, one per stored panel component), sorted. */
export function pluginPanelGroupPaths(node: SceneNode): string[] {
  const out: string[] = [];
  for (const c of node.components) {
    const p = parsePanelType(c.type);
    if (p) out.push(`plugin/${p.slug}/${p.panel}`);
  }
  return out.sort();
}

// ── Fields (non-numeric values) ───────────────────────────────────────

export function readPluginField(node: SceneNode, b: PropBinding): Value {
  const comp = node.components.find((c) => c.type === b.field!.groupId);
  const v = (comp?.props as Record<string, unknown> | undefined)?.[b.field!.key];
  if (b.valueType === 'bool') return { kind: 'bool', value: v === true };
  if (b.valueType === 'string') return { kind: 'string', value: typeof v === 'string' ? v : '' };
  return { kind: 'json', value: JSON.stringify(v === undefined ? null : v) };
}

/** The raw value a plugin field stores for `value` (its own type, or ANY json). */
function pluginRaw(b: PropBinding, value: Value): unknown {
  switch (value.kind) {
    case 'bool': if (b.valueType === 'bool') return value.value; break;
    case 'string': if (b.valueType === 'string') return value.value; break;
    case 'json':
      try { return JSON.parse(value.value) as unknown; } catch { return fail('invalidArgument', 'invalid json', { path: b.path }); }
    default: break;
  }
  return fail('typeMismatch', `'${b.path}' takes a ${b.valueType} (or json), got ${value.kind}`, { path: b.path, detail: JSON.stringify({ expected: b.valueType }) });
}

export function writePluginField(layerId: string, node: SceneNode, b: PropBinding, value: Value): void {
  const raw = pluginRaw(b, value);
  const comp = node.components.find((c) => c.type === b.field!.groupId);
  if (!comp) fail('notFound', `layer '${layerId}' has no ${b.field!.groupId}`, { layer: layerId, path: b.path });
  // Stored verbatim — JSON null included (an empty asset slot): deleting the key
  // would remove the property.
  defaultSceneGraph.writeProp(layerId, comp.id, b.field!.key, raw);
}

// ── Panel groups (addPropertyGroup / removePropertyGroups) ───────────

/** `plugin/<slug>/<panel>` → the panel it names, or null (not a panel group path). */
export function parsePanelGroupPath(path: string): { slug: string; panel: string; type: string; id: string } | null {
  const seg = path.split('/');
  if (seg.length !== 3 || seg[0] !== 'plugin' || !SEG.test(seg[1]!) || !SEG.test(seg[2]!)) return null;
  return {
    slug: seg[1]!, panel: seg[2]!,
    type: `${PLUGIN_PANEL_COMPONENT_PREFIX}${seg[1]}.${seg[2]}`,
    id: `${PLUGIN_PANEL_ID_PREFIX}${seg[1]}_${seg[2]}`,
  };
}

/** The group a `PluginParams.<slug>.<panel>` match name adds under `plugin`, or null. */
export function panelGroupForMatchName(matchName: string): string | null {
  const p = parsePanelType(matchName);
  return p ? `plugin/${p.slug}/${p.panel}` : null;
}

/** An `init` value as the stored props (a point = `<name>.x` / `.y` / `.z` numbers). */
export function panelInitProps(init: ReadonlyArray<{ path: string; value: Value }>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const { path, value } of init) {
    if (!/^[a-z][a-zA-Z0-9]{0,31}$/.test(path)) fail('invalidArgument', `'${path}' is not a plugin param name`, { path });
    const num = (x: number): number => {
      if (!Number.isFinite(x)) fail('invalidArgument', `'${path}': value must be finite`, { path });
      return x;
    };
    switch (value.kind) {
      case 'scalar': case 'int': out[path] = num(value.value); break;
      case 'bool': out[path] = value.value; break;
      case 'string': case 'choice': out[path] = value.value; break;
      case 'vec2': out[`${path}.x`] = num(value.value.x); out[`${path}.y`] = num(value.value.y); break;
      case 'vec3': out[`${path}.x`] = num(value.value.x); out[`${path}.y`] = num(value.value.y); out[`${path}.z`] = num(value.value.z); break;
      case 'color': out[path] = channelsToColor(value.value.r, value.value.g, value.value.b, value.value.a); break;
      case 'json':
        try { out[path] = JSON.parse(value.value) as unknown; } catch { fail('invalidArgument', 'invalid json', { path }); }
        break;
      default: fail('typeMismatch', `'${path}': a plugin param takes a number, point, bool, string, colour or json`, { path });
    }
  }
  return out;
}

File: native/engine/src/core/plugin_props.cpp
#include "plugin_props.hpp"

#include <algorithm>
#include <array>
#include <cmath>
#include <map>
#include <set>

#include "fail.hpp"
#include "fxstate.hpp"
#include "scene.hpp"
#include "strutil.hpp"

namespace premation::doc {
namespace {

using api::ErrorCode;
using api::ValueType;
using js::stringify;

constexpr std::string_view kLayerComponentPrefix = "pluginLayer:";
constexpr std::string_view kLayerTrackPrefix = "plugin.";
constexpr std::string_view kPanelComponentPrefix = "PluginParams.";
constexpr std::string_view kPanelIdPrefix = "pluginui_";
constexpr std::array<std::string_view, 3> kAxes{"x", "y", "z"};

/// pluginProps.ts SEG: /^[A-Za-z0-9-]+$/.
bool seg_ok(std::string_view s) {
  if (s.empty()) return false;
  return std::all_of(s.begin(), s.end(), [](char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '-';
  });
}

bool is_axis(std::string_view a) { return std::find(kAxes.begin(), kAxes.end(), a) != kAxes.end(); }

std::vector<std::string> split_dots(std::string_view s) {
  std::vector<std::string> out;
  std::size_t start = 0;
  while (true) {
    const std::size_t dot = s.find('.', start);
    if (dot == std::string_view::npos) {
      out.emplace_back(s.substr(start));
      return out;
    }
    out.emplace_back(s.substr(start, dot - start));
    start = dot + 1;
  }
}

struct PluginTrack {
  bool layer = false;
  std::string slug;
  std::string panel;
  std::string name;
  std::optional<std::string> axis;
};

/// `parsePluginTrack(prop)`.
std::optional<PluginTrack> parse_plugin_track(std::string_view prop) {
  if (prop.starts_with(kPluginPanelTrackPrefix)) {
    const std::vector<std::string> seg = split_dots(prop.substr(kPluginPanelTrackPrefix.size()));
    if (seg.size() < 3 || seg.size() > 4) return std::nullopt;
    if (!std::all_of(seg.begin(), seg.end(), [](const std::string& s) { return seg_ok(s); })) return std::nullopt;
    PluginTrack t;
    t.slug = seg[0];
    t.panel = seg[1];
    t.name = seg[2];
    if (seg.size() == 4) {
      if (!is_axis(seg[3])) return std::nullopt;
      t.axis = seg[3];
    }
    return t;
  }
  if (prop.starts_with(kLayerTrackPrefix)) {
    const std::string_view name = prop.substr(kLayerTrackPrefix.size());
    if (!seg_ok(name) || name.starts_with('_')) return std::nullopt;
    PluginTrack t;
    t.layer = true;
    t.name = std::string(name);
    return t;
  }
  return std::nullopt;
}

const Component* layer_kind_component(const Node& n) {
  for (const Component& c : n.components) {
    if (c.type.starts_with(kLayerComponentPrefix)) return &c;
  }
  return nullptr;
}

std::string panel_type(std::string_view slug, std::string_view panel) {
  return std::string(kPanelComponentPrefix) + std::string(slug) + "." + std::string(panel);
}

/// `parsePanelType(type)`.
std::optional<std::pair<std::string, std::string>> parse_panel_type(std::string_view type) {
  if (!type.starts_with(kPanelComponentPrefix)) return std::nullopt;
  const std::vector<std::string> seg = split_dots(type.substr(kPanelComponentPrefix.size()));
  if (seg.size() != 2 || !seg_ok(seg[0]) || !seg_ok(seg[1])) return std::nullopt;
  return std::make_pair(seg[0], seg[1]);
}

/// pluginProps.ts `ownKey`.
bool own_key(std::string_view k) {
  if (k.starts_with('_')) return false;
  const std::vector<std::string> seg = split_dots(k);
  return seg_ok(seg[0]) && seg.size() <= 2;
}

struct Home {
  const Component* comp = nullptr;
  std::string key;
};

Home static_home(const Node& n, const PluginTrack& t) {
  if (t.layer) return Home{layer_kind_component(n), t.name};
  return Home{n.comp(panel_type(t.slug, t.panel)), t.axis ? t.name + "." + *t.axis : t.name};
}

ValueType field_type(const Json& v) {
  if (v.is_bool()) return ValueType::bool_;
  if (v.is_string()) return ValueType::string;
  return ValueType::json;
}

PropBinding numeric_binding(std::string path, const std::string& name, std::vector<std::string> members) {
  PropBinding b;
  b.path = std::move(path);
  b.name = name;
  b.matchName = members[0];
  b.valueType = members.size() <= 1 ? ValueType::scalar : members.size() == 2 ? ValueType::vec2 : ValueType::vec3;
  b.members = std::move(members);
  b.animatable = true;
  // The static seam owns the value: nothing is invented on the Transform.
  b.home = std::vector<std::string>{};
  return b;
}

PropBinding field_binding(std::string path, const std::string& name, const std::string& component, const std::string& key,
                          const Json& stored) {
  PropBinding b;
  b.path = std::move(path);
  b.name = name;
  b.matchName = key;
  b.valueType = field_type(stored);
  b.special = Special::field;
  b.field = FieldRef{"plugin", key, component, std::nullopt};
  b.animatable = false;
  return b;
}

}  // namespace

std::optional<std::string> plugin_api_path(std::string_view prop) {
  const auto t = parse_plugin_track(prop);
  if (!t) return std::nullopt;
  if (t->layer) return "plugin/" + t->name;
  return "plugin/" + t->slug + "/" + t->panel + "/" + t->name;
}

std::optional<std::optional<double>> read_plugin_static(const Node& n, std::string_view prop) {
  const auto t = parse_plugin_track(prop);
  if (!t) return std::nullopt;
  const Home h = static_home(n, *t);
  if (h.comp == nullptr) return std::optional<double>{};
  const Json& v = h.comp->props.at(h.key);
  return v.is_number() ? std::optional<double>(v.num()) : std::optional<double>{};
}

std::optional<bool> write_plugin_static(Document& d, std::string_view nodeId, std::string_view prop, double value) {
  const auto t = parse_plugin_track(prop);
  if (!t) return std::nullopt;
  const Node* n = d.node(nodeId);
  if (n == nullptr) return false;
  const Home h = static_home(*n, *t);
  if (h.comp == nullptr) return false;
  const std::string cid = h.comp->id;
  return sg_write_prop(d, nodeId, cid, h.key, Json::number(value));
}

void add_plugin_bindings(const Node& node, const std::vector<std::string>& trackNames,
                         const std::function<void(PropBinding)>& add) {
  std::vector<PluginTrack> tracks;
  for (const auto& p : trackNames) {
    if (auto t = parse_plugin_track(p)) tracks.push_back(std::move(*t));
  }
  if (const Component* kind = layer_kind_component(node)) {
    std::set<std::string> names;  // sorted, like JS [...names].sort() (ASCII)
    for (const auto& m : kind->props.obj()) {
      if (own_key(m.key) && m.key.find('.') == std::string::npos) names.insert(m.key);
    }
    for (const auto& t : tracks) {
      if (t.layer) names.insert(t.name);
    }
    for (const auto& name : names) {
      const Json& v = kind->props.at(name);
      std::string path = "plugin/" + name;
      if (v.is_undefined() || v.is_number()) add(numeric_binding(std::move(path), name, {std::string(kLayerTrackPrefix) + name}));
      else add(field_binding(std::move(path), name, kind->type, name, v));
    }
  }
  struct Panel {
    std::string slug;
    std::string panel;
    const Component* comp = nullptr;
  };
  std::map<std::string, Panel> panels;  // sorted by component type
  for (const Component& c : node.components) {
    if (auto p = parse_panel_type(c.type)) panels.insert_or_assign(c.type, Panel{p->first, p->second, &c});
  }
  for (const auto& t : tracks) {
    if (t.layer) continue;
    const std::string type = panel_type(t.slug, t.panel);
    if (!panels.contains(type)) panels.emplace(type, Panel{t.slug, t.panel, nullptr});
  }
  for (const auto& [type, p] : panels) {
    std::map<std::string, std::set<std::string>> numeric;
    std::map<std::string, Json> other;
    if (p.comp != nullptr) {
      for (const auto& m : p.comp->props.obj()) {
        if (!own_key(m.key) || m.value.is_undefined()) continue;
        const std::vector<std::string> seg = split_dots(m.key);
        const std::string& name = seg[0];
        if (m.value.is_number()) {
          if (seg.size() == 2 && !is_axis(seg[1])) continue;
          numeric[name].insert(seg.size() == 2 ? seg[1] : std::string());
        } else if (seg.size() == 1) {
          other.insert_or_assign(name, m.value);
        }
      }
    }
    for (const auto& t : tracks) {
      if (t.layer || t.slug != p.slug || t.panel != p.panel || other.contains(t.name)) continue;
      numeric[t.name].insert(t.axis.value_or(std::string()));
    }
    std::set<std::string> names;
    for (const auto& [n, _] : numeric) names.insert(n);
    for (const auto& [n, _] : other) names.insert(n);
    const std::string base = "plugin/" + p.slug + "/" + p.panel;
    const std::string prefix = std::string(kPluginPanelTrackPrefix) + p.slug + "." + p.panel + ".";
    for (const auto& name : names) {
      std::string path = base + "/" + name;
      if (const auto it = other.find(name); it != other.end()) {
        add(field_binding(std::move(path), name, type, name, it->second));
        continue;
      }
      std::vector<std::string> members;
      for (const auto a : kAxes) {
        if (numeric[name].contains(std::string(a))) members.push_back(prefix + name + "." + std::string(a));
      }
      if (members.empty()) members.push_back(prefix + name);
      add(numeric_binding(std::move(path), name, std::move(members)));
    }
  }
}

std::vector<std::string> plugin_panel_group_paths(const Node& node) {
  std::vector<std::string> out;
  for (const Component& c : node.components) {
    if (auto p = parse_panel_type(c.type)) out.push_back("plugin/" + p->first + "/" + p->second);
  }
  std::sort(out.begin(), out.end());
  return out;
}

api::Value read_plugin_field(const Node& node, const PropBinding& b) {
  const Component* comp = node.comp(b.field->animatorId.value_or(""));
  const Json v = comp != nullptr ? comp->props.at(b.field->key) : Json();
  if (b.valueType == ValueType::bool_) return v_bool(v.is_bool() && v.b());
  if (b.valueType == ValueType::string) return v_string(v.is_string() ? v.str() : std::string());
  return v_json(stringify(v.is_undefined() ? Json::null() : v));
}

void write_plugin_field(Document& d, std::string_view layer, const PropBinding& b, const api::Value& value) {
  Json raw;
  if (value.kind() == VK::bool_ && b.valueType == ValueType::bool_) {
    raw = Json::boolean(get<VK::bool_>(value));
  } else if (value.kind() == VK::string && b.valueType == ValueType::string) {
    raw = Json::string(get<VK::string>(value));
  } else if (value.kind() == VK::json) {
    auto parsed = js::parse(get<VK::json>(value));
    if (!parsed) fail(ErrorCode::invalid_argument, "invalid json", {.path = b.path});
    raw = std::move(*parsed);
  } else {
    fail(ErrorCode::type_mismatch,
         "'" + b.path + "' takes a " + std::string(value_type_name(b.valueType)) + " (or json), got " +
             std::string(kind_name(value.kind())),
         {.path = b.path, .detail = "{\"expected\":\"" + std::string(value_type_name(b.valueType)) + "\"}"});
  }
  const std::string type = b.field->animatorId.value_or("");
  const Node* n = d.node(layer);
  const Component* comp = n != nullptr ? n->comp(type) : nullptr;
  if (comp == nullptr) {
    fail(ErrorCode::not_found, "layer '" + std::string(layer) + "' has no " + type, {.layer = std::string(layer), .path = b.path});
  }
  const std::string cid = comp->id;
  // Stored verbatim — JSON null included (an empty asset slot).
  (void)sg_write_prop(d, layer, cid, b.field->key, std::move(raw));
}

std::optional<PanelGroup> parse_panel_group_path(std::string_view path) {
  const std::vector<std::string> seg = [&] {
    std::vector<std::string> out;
    std::size_t start = 0;
    while (true) {
      const std::size_t slash = path.find('/', start);
      if (slash == std::string_view::npos) {
        out.emplace_back(path.substr(start));
        return out;
      }
      out.emplace_back(path.substr(start, slash - start));
      start = slash + 1;
    }
  }();
  if (seg.size() != 3 || seg[0] != "plugin" || !seg_ok(seg[1]) || !seg_ok(seg[2])) return std::nullopt;
  return PanelGroup{seg[1], seg[2], panel_type(seg[1], seg[2]), std::string(kPanelIdPrefix) + seg[1] + "_" + seg[2]};
}

std::optional<std::string> panel_group_for_match_name(std::string_view matchName) {
  const auto p = parse_panel_type(matchName);
  if (!p) return std::nullopt;
  return "plugin/" + p->first + "/" + p->second;
}

Json panel_init_props(const std::vector<api::PropertyInit>& init) {
  Json out = Json::object();
  for (const auto& i : init) {
    const std::string& path = i.path;
    // /^[a-z][a-zA-Z0-9]{0,31}$/
    const bool nameOk = !path.empty() && path.size() <= 32 && path[0] >= 'a' && path[0] <= 'z' &&
                        std::all_of(path.begin(), path.end(), [](char c) {
                          return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9');
                        });
    if (!nameOk) fail(ErrorCode::invalid_argument, "'" + path + "' is not a plugin param name", {.path = path});
    auto num = [&](double x) {
      if (!std::isfinite(x)) fail(ErrorCode::invalid_argument, "'" + path + "': value must be finite", {.path = path});
      return Json::number(x);
    };
    const api::Value& v = i.value;
    switch (v.kind()) {
      case VK::scalar: out.set(path, num(get<VK::scalar>(v))); break;
      case VK::int_: out.set(path, num(static_cast<double>(get<VK::int_>(v)))); break;
      case VK::bool_: out.set(path, Json::boolean(get<VK::bool_>(v))); break;
      case VK::string: out.set(path, Json::string(get<VK::string>(v))); break;
      case VK::choice: out.set(path, Json::string(get<VK::choice>(v))); break;
      case VK::vec2:
        out.set(path + ".x", num(get<VK::vec2>(v).x));
        out.set(path + ".y", num(get<VK::vec2>(v).y));
        break;
      case VK::vec3:
        out.set(path + ".x", num(get<VK::vec3>(v).x));
        out.set(path + ".y", num(get<VK::vec3>(v).y));
        out.set(path + ".z", num(get<VK::vec3>(v).z));
        break;
      case VK::color: {
        const api::Color& c = get<VK::color>(v);
        out.set(path, Json::string(channels_to_color(c.r, c.g, c.b, c.a)));
        break;
      }
      case VK::json: {
        auto parsed = js::parse(get<VK::json>(v));
        if (!parsed) fail(ErrorCode::invalid_argument, "invalid json", {.path = path});
        out.set(path, std::move(*parsed));
        break;
      }
      default:
        fail(ErrorCode::type_mismatch, "'" + path + "': a plugin param takes a number, point, bool, string, colour or json",
             {.path = path});
    }
  }
  return out;
}

}  // namespace premation::doc
